teuthology/suite/run.py

```diff
@@ -27,6 +27,45 @@
 
 log = logging.getLogger(__name__)
 
+def descr_to_yamls(descriptions, suites_path):
+    """
+    Function converts description of a job into sequence of .yaml files.
+    Input "descriptions" is a string containing comma-separated list of job descriptions.
+    Input "suites_path" is a posix path to dir containing all suites.
+    """
+    def expand_descr(test_yamls, source, prefix, base_pos):
+        """
+        Expand description using production rules (explanation, not Chomsky context-free formalism):
+        rule 1: (simplification)
+        "A{BX}" => AB A{X}
+        rule 2: (termination)
+        "A" => "suites_pathA.yaml"
+        """
+        pos = base_pos
+        while (pos < len(source)):
+            if source[pos] == '{':
+                more_prefix=source[base_pos:pos]
+                pos = expand_descr(test_yamls, source, prefix + more_prefix, pos + 1)
+                base_pos = pos
+            elif source[pos] == '}':
+                if base_pos != pos:
+                    test_yamls.append(suites_path + "/" + prefix + source[base_pos:pos] + ".yaml")
+                return pos + 1
+            elif source[pos] == ' ':
+                if base_pos != pos:
+                    test_yamls.append(suites_path + "/" + prefix + source[base_pos:pos] + ".yaml")
+                pos = pos + 1
+                base_pos = pos
+            else:
+                pos = pos + 1
+    result = []
+    desc_tab = descriptions.split(',')
+    for d in desc_tab:
+        dd = d.strip()
```

Comment:
instead of creating another variable you could use python style generator expression:
```
for d in (_.strip() for _ in descriptions.split(',')):
```

Reply:
I apologise for the next example, but I was not able to refrain. As far as I understood this part of the code I would refactor it like that:
```
    def expand_path(paths: list[str], source: str, prefix: str, pos: int) -> int:
        """
        Expand description using production rules (explanation, not Chomsky context-free formalism):
        1) simplification: "A{BX}" => AB A{X}
        2) termination:    "A" => "A"
            
        :param paths:   pointer to the extendible list of yaml paths expanded from the source string
        :param source:  source string containing a single test description
        :param prefix:  suffix a string representation of a part of path
        :param pos:     base position in the source string to scan next portion of prefix
            
        :returns:       index of a character following closed brace, i.e. character "}".
        """
        p = i = pos
        while (i < len(source)):
            path = prefix + source[p:i]
            match (source[i]):
                case '{':
                    p = i = expand_path(paths, source, path, i + 1)
                    continue
                case '}':
                    if p != i:
                        paths.append(path)
                    return i + 1
                case ' ':
                    if p != i:
                        paths.append(path)
                    p = i = i + 1
                case _:
                    i = i + 1

    for d in (_.strip() for _ in descriptions.split(',')):
        yamls = []
        expand_path(yamls, d, "", 0)
        yield (d, [f"{base_path}/{_}.yaml" for _ in yamls])

```